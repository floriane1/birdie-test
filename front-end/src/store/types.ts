export interface Event {
  id: string;
  event_type: EventType;
  care_recipient_id: string;
  timestamp: Date;
  visit_id?: string;
  caregiver_id?: string;
  observation_event_id?: string;
  // tslint:disable-next-line:no-any
  [key: string]: any;
}

export enum EventType {
  ALERT_QUALIFIED = 'alert_qualified',
  ALERT_RAISED = 'alert_raised',
  CATHETER_OBSERVATION = 'catheter_observation',
  CHECK_IN = 'check_in',
  CHECK_OUT = 'check_out',
  CONCERN_RAISED = 'concern_raised',
  FLUID_INTAKE_OBSERVATION = 'fluid_intake_observation',
  FOOD_INTAKE_OBSERVATION = 'food_intake_observation',
  GENERAL_OBSERVATION = 'general_observation',
  INCONTINENCE_PAD_OBSERVATION = 'incontinence_pad_observation',
  MEDICATION_SCHEDULE_CREATED = 'medication_schedule_created',
  MEDICATION_SCHEDULE_UPDATED = 'medication_schedule_updated',
  MENTAL_HEALTH_OBSERVATION = 'mental_health_observation',
  MOOD_OBSERVATION = 'mood_observation',
  NO_MEDICATION_OBSERVATION_RECEIVED = 'no_medication_observation_received',
  PHYSICAL_HEALTH_OBSERVATION = 'physical_health_observation',
  REGULAR_MEDICATION_MAYBE_TAKEN = 'regular_medication_maybe_taken',
  REGULAR_MEDICATION_NOT_TAKEN = 'regular_medication_not_taken',
  REGULAR_MEDICATION_PARTIALLY_TAKEN = 'regular_medication_partially_taken',
  REGULAR_MEDICATION_TAKEN = 'regular_medication_taken',
  TASK_COMPLETED = 'task_completed',
  TASK_COMPLETION_REVERTED = 'task_completion_reverted',
  TASK_SCHEDULE_CREATED = 'task_schedule_created',
  TOILET_VISIT_RECORDED = 'toilet_visit_recorded',
  VISIT_CANCELLED = 'visit_cancelled',
  VISIT_COMPLETED = 'visit_completed',
}
