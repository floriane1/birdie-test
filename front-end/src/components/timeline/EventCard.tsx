import * as React from 'react';
import {
  ExclamationTriangle,
  Syringe,
  Utensils,
  NotesMedical,
  Toilet,
  Pills,
  CalendarCheck,
  CalendarMinus,
  CalendarPlus,
  CalendarTimes,
  Smile,
  SignInAlt,
  SignOutAlt,
  Meh,
  MehBlank,
  Frown,
} from '@styled-icons/fa-solid';
import { Body } from '@styled-icons/ionicons-outline';
import { LocalDrink, Schedule } from '@styled-icons/material';
import { MentalHealth } from '@styled-icons/remix-fill';
import { StyledIcon } from '@styled-icons/styled-icon';
import * as dayjs from 'dayjs';
import * as relativeTime from 'dayjs/plugin/relativeTime';
dayjs.extend(relativeTime);

import { Event, EventType } from '@App/store/types.ts';
import { VerticalTimelineElement } from 'react-vertical-timeline-component';

interface EventCardProps {
  event: Event;
}

const colors = {
  red: '#ff0000',
  green: '#39D576',
  orange: '#ffa500',
  grey: '#a9a9a9',
  white: '#ffffff',
};

interface EventAttributes {
  color: string;
  icon: StyledIcon;
}

const eventAttributes = (event: Event): EventAttributes => {
  const { event_type } = event;
  switch (event_type) {
    case EventType.ALERT_QUALIFIED:
      return { color: colors.red, icon: ExclamationTriangle };
    case EventType.ALERT_RAISED:
      return { color: colors.red, icon: ExclamationTriangle };
    case EventType.CATHETER_OBSERVATION:
      return { color: colors.grey, icon: Syringe };
    case EventType.CHECK_IN:
      return { color: colors.green, icon: SignInAlt };
    case EventType.CHECK_OUT:
      return { color: colors.green, icon: SignOutAlt };
    case EventType.CONCERN_RAISED:
      return { color: colors.red, icon: ExclamationTriangle };
    case EventType.FLUID_INTAKE_OBSERVATION:
      return { color: colors.grey, icon: LocalDrink };
    case EventType.FOOD_INTAKE_OBSERVATION:
      return { color: colors.grey, icon: Utensils };
    case EventType.GENERAL_OBSERVATION:
      return { color: colors.grey, icon: NotesMedical };
    case EventType.INCONTINENCE_PAD_OBSERVATION:
      return { color: colors.grey, icon: Toilet };
    case EventType.MEDICATION_SCHEDULE_CREATED:
      return { color: colors.green, icon: Schedule };
    case EventType.MEDICATION_SCHEDULE_UPDATED:
      return { color: colors.green, icon: Schedule };
    case EventType.MENTAL_HEALTH_OBSERVATION:
      return { color: colors.grey, icon: MentalHealth };
    case EventType.NO_MEDICATION_OBSERVATION_RECEIVED:
      return { color: colors.red, icon: NotesMedical };
    case EventType.PHYSICAL_HEALTH_OBSERVATION:
      return { color: colors.grey, icon: Body };
    case EventType.REGULAR_MEDICATION_MAYBE_TAKEN:
      return { color: colors.orange, icon: Pills };
    case EventType.REGULAR_MEDICATION_NOT_TAKEN:
      return { color: colors.red, icon: Pills };
    case EventType.REGULAR_MEDICATION_PARTIALLY_TAKEN:
      return { color: colors.orange, icon: Pills };
    case EventType.REGULAR_MEDICATION_TAKEN:
      return { color: colors.green, icon: Pills };
    case EventType.TASK_COMPLETED:
      return { color: colors.green, icon: CalendarCheck };
    case EventType.TASK_COMPLETION_REVERTED:
      return { color: colors.red, icon: CalendarMinus };
    case EventType.TASK_SCHEDULE_CREATED:
      return { color: colors.green, icon: CalendarPlus };
    case EventType.TOILET_VISIT_RECORDED:
      return { color: colors.grey, icon: Toilet };
    case EventType.VISIT_CANCELLED:
      return { color: colors.red, icon: CalendarTimes };
    case EventType.VISIT_COMPLETED:
      return { color: colors.green, icon: CalendarCheck };
    case EventType.MOOD_OBSERVATION:
      const moodIcon = (mood: string) => {
        switch (mood) {
          case 'happy':
            return Smile;
          case 'okay':
            return Meh;
          case 'sad':
            return Frown;
          default:
            return MehBlank;
        }
      };
      const moodColor = (mood: string) => {
        switch (mood) {
          case 'happy':
            return colors.green;
          case 'okay':
            return colors.orange;
          case 'sad':
            return colors.red;
          default:
            return colors.grey;
        }
      };
      return { color: moodColor(event.mood), icon: moodIcon(event.mood) };
    default:
      return { color: colors.grey, icon: NotesMedical };
  }
};

export const EventCard = ({ event }: EventCardProps) => {
  const { id, event_type, timestamp } = event;
  const eventTitle: string =
    event_type.charAt(0).toUpperCase() + event_type.slice(1).replace(/_/g, ' ');
  const Icon = eventAttributes(event).icon;
  return (
    <VerticalTimelineElement
      date={dayjs(timestamp).format('dddd, MMMM D, h:mm A')}
      iconStyle={{
        color: colors.white,
        background: eventAttributes(event).color,
      }}
      icon={<Icon />}
      contentStyle={{
        padding: '0 16px',
        borderTop: `6px solid ${eventAttributes(event).color}`,
      }}
      key={id}
    >
      <h3 style={{ marginBottom: 0 }}>{eventTitle}</h3>
      {event.note && <p>Note: {event.note}</p>}
      {event.task_schedule_note && <p>Note: {event.task_schedule_note}</p>}
      {event.task_definition_description && (
        <p>Description: {event.task_definition_description}</p>
      )}
      {event.volume_ml && <p>Volume: {event.volume_ml}mL</p>}
      {event.meal && <p>Meal: {event.meal}</p>}
      {event.fluid && <p>Type of fluid: {event.fluid}</p>}
      {event.observed && <p>Observed: {event.observed ? 'yes' : 'no'}</p>}
      {event.medication_failure_reason && (
        <p>Reason: {event.medication_failure_reason}</p>
      )}
      {event.severity && <p>Severity: {event.severity}</p>}
      {event.alert_severity && <p>Severity: {event.alert_severity}</p>}
      {event.consumed_volume_ml && <p>Volume: {event.consumed_volume_ml}mL</p>}
      {event.pad_condition && <p>Condition: {event.pad_condition}</p>}
    </VerticalTimelineElement>
  );
};
