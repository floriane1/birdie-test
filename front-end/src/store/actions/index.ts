import { Event } from '../types';
import {
  FetchEventsRequestAction,
  FetchEventsSuccessAction,
  FetchEventsErrorAction,
  FETCH_EVENTS_REQUEST,
  FETCH_EVENTS_SUCCESS,
  FETCH_EVENTS_ERROR,
} from '../types/index';

export function fetchEventsRequest(
  careRecipientId: string
): FetchEventsRequestAction {
  return {
    type: FETCH_EVENTS_REQUEST,
    payload: careRecipientId,
  };
}

export function fetchEventsSuccess(payload: Event[]): FetchEventsSuccessAction {
  return {
    type: FETCH_EVENTS_SUCCESS,
    payload,
  };
}

export function fetchEventsError(): FetchEventsErrorAction {
  return {
    type: FETCH_EVENTS_ERROR,
  };
}
