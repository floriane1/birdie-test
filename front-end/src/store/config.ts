export const CARE_RECIPIENT_DEFAULT_ID: string =
  'ad3512a6-91b1-4d7d-a005-6f8764dd0111';
