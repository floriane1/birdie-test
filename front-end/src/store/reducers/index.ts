import { combineReducers } from 'redux';
import { Event } from '../types';
import {
  FetchEventsAction,
  FETCH_EVENTS_REQUEST,
  FETCH_EVENTS_SUCCESS,
  FETCH_EVENTS_ERROR,
} from '../types/index';

export type RootState = Readonly<{
  events: Event[];
  loading: boolean;
  error: boolean;
}>;

export const initialState: RootState = {
  events: [],
  loading: false,
  error: false,
};

export const eventsReducer = (
  state: RootState = initialState,
  action: FetchEventsAction
): RootState => {
  switch (action.type) {
    case FETCH_EVENTS_REQUEST:
      return { ...state, loading: true, error: false };
    case FETCH_EVENTS_SUCCESS:
      return { ...state, events: action.payload, loading: false, error: false };
    case FETCH_EVENTS_ERROR:
      return { ...state, loading: false, error: true };
    default:
      return state;
  }
};

export const rootReducer = combineReducers<RootState>({ eventsReducer });
