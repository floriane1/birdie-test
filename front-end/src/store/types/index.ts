import { Event } from '../types';

export const FETCH_EVENTS_REQUEST = 'FETCH_EVENTS_REQUEST';
export const FETCH_EVENTS_SUCCESS = 'FETCH_EVENTS_SUCCESS';
export const FETCH_EVENTS_ERROR = 'FETCH_EVENTS_ERROR';

export interface FetchEventsRequestAction {
  type: typeof FETCH_EVENTS_REQUEST;
  payload: string;
}

export interface FetchEventsSuccessAction {
  type: typeof FETCH_EVENTS_SUCCESS;
  payload: Event[];
}

export interface FetchEventsErrorAction {
  type: typeof FETCH_EVENTS_ERROR;
  payload: string;
}

export type FetchEventsAction =
  | FetchEventsRequestAction
  | FetchEventsSuccessAction
  | FetchEventsErrorAction;
