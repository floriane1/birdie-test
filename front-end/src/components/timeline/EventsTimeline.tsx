import * as React from 'react';
import styled from 'styled-components';
import { Event } from '@App/store/types.ts';
import { EventCard } from '@App/components/timeline/EventCard.tsx';

import { VerticalTimeline } from 'react-vertical-timeline-component';
import 'react-vertical-timeline-component/style.min.css';

type EventsTimelineProps = {
  events: Event[];
};

const TimelineContainer = styled.div`
  height: 65%;
  width: 85%;
  position: relative;
`;

export const EventsTimeline = ({ events }: EventsTimelineProps) => {
  return (
    <TimelineContainer>
      {events.length > 0 && (
        <VerticalTimeline layout="1-column">
          {events.map((event) => (
            <EventCard key={event.id} event={event} />
          ))}
          ;
        </VerticalTimeline>
      )}
    </TimelineContainer>
  );
};
