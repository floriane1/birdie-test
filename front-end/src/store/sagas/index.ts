import { call, put, all, takeLatest } from 'redux-saga/effects';
import axios from 'axios';
import { FetchEventsRequestAction, FETCH_EVENTS_REQUEST } from '../types/index';
import { fetchEventsSuccess, fetchEventsError } from '../actions';
import { Event } from '../types';

export function* fetchEvents(action: FetchEventsRequestAction) {
  try {
    const query = `/care-recipients/${action.payload}/events`;
    const results = yield call(axios.get, query);
    const events: Event[] = results.data.map((event: Event) => ({
      ...event,
      timestamp: new Date(
        new Date(event.timestamp).setMonth(
          // TO-DO: change when events data more recent
          new Date(event.timestamp).getMonth() + 17
        )
      ),
    }));
    yield put(fetchEventsSuccess(events));
  } catch (error) {
    yield put(fetchEventsError(error.message));
  }
}

export function* watchFetchEvents() {
  yield takeLatest(FETCH_EVENTS_REQUEST, fetchEvents);
}

function* initSaga() {
  yield all([watchFetchEvents()]);
}

export default initSaga;
