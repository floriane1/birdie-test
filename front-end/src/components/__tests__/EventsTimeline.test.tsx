import * as React from 'react';
import { shallow, configure } from 'enzyme';
import * as Adapter from 'enzyme-adapter-react-16';
import { EventsTimeline } from '@App/components/timeline/EventsTimeline';
import { Event, EventType } from '@App/store//types';

const eventTestA: Event = {
  id: 'event-id-A',
  event_type: EventType.ALERT_RAISED,
  care_recipient_id: 'care-recipient-id-123',
  timestamp: new Date('2019-04-24T02:00:11.187Z'),
};

const eventTestB: Event = {
    id: 'event-id-B',
    event_type: EventType.ALERT_RAISED,
    care_recipient_id: 'care-recipient-id-123',
    timestamp: new Date('2019-05-24T02:00:11.187Z'),
  };

configure({ adapter: new Adapter() });

describe('EventsTimeline', () => {
  it('renders correctly', () => {
    const wrapper = shallow(<EventsTimeline events={[eventTestA, eventTestB]} />);
    expect(wrapper).toMatchSnapshot();
  });
});
