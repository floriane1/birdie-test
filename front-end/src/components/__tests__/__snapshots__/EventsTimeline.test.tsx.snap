// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventsTimeline renders correctly 1`] = `
<styled.div>
  <VerticalTimeline
    animate={true}
    className=""
    layout="1-column"
  >
    <Component
      event={
        Object {
          "care_recipient_id": "care-recipient-id-123",
          "event_type": "alert_raised",
          "id": "event-id-A",
          "timestamp": 2019-04-24T02:00:11.187Z,
        }
      }
      key="event-id-A"
    />
    <Component
      event={
        Object {
          "care_recipient_id": "care-recipient-id-123",
          "event_type": "alert_raised",
          "id": "event-id-B",
          "timestamp": 2019-05-24T02:00:11.187Z,
        }
      }
      key="event-id-B"
    />
  </VerticalTimeline>
</styled.div>
`;
